src/dbManager.js

```diff
@@ -0,0 +1,161 @@
+const BN = require('bn.js')
+const level = require('level-mem')
+const rlp = require('rlp')
+const Block = require('ethereumjs-block')
+const Cache = require('./cache')
+const {
+  headsKey,
+  headHeaderKey,
+  headBlockKey,
+  hashToNumberKey,
+  numberToHashKey,
+  tdKey,
+  bodyKey,
+  headerKey
+} = require('./util')
+
+module.exports = class DBManager {
+  constructor (db, common) {
+    this._db = db
+    this._common = common
+    this._opts = {
+      keyEncoding: 'binary',
+      valueEncoding: 'binary'
+    }
+    this._cache = {
+      td: new Cache({ max: 1024 }),
+      header: new Cache({ max: 512 }),
+      body: new Cache({ max: 256 }),
+      numberToHash: new Cache({ max: 2048 }),
+      hashToNumber: new Cache({ max: 2048 })
+    }
+  }
+
+  getHeads () {
+    return this._db.get(headsKey, {
+      keyEncoding: 'binary',
+      valueEncoding: 'json'
+    })
+  }
+
+  getHeadHeader () {
+    return this._db.get(headHeaderKey, this._opts)
+  }
+
+  getHeadBlock () {
+    return this._db.get(headBlockKey, this._opts)
+  }
+
+  get (k) {
+    return this._db.get(k, this._opts)
+  }
+
+  batch (ops) {
+    return this._db.batch(ops)
+  }
+
+  /**
+   * Performs a block hash to block number lookup
+   * @method _hashToNumber
+   */
+  async _hashToNumber (hash) {
+    const key = hashToNumberKey(hash)
+    let number = this._cache.hashToNumber.get(key)
+    if (number) {
+      return new BN(number)
+    }
+
+    number = await this.get(key)
+    this._cache.hashToNumber.set(key, number)
+
+    return new BN(number)
+  }
+
+  /**
+   * Performs a block number to block hash lookup
+   * @method _numberToHash
+   */
+  async _numberToHash (number) {
+    if (number.ltn(0)) {
+      throw new level.errors.NotFoundError()
+    }
+
+    const key = numberToHashKey(number)
+    let hash = this._cache.numberToHash.get(key)
+    if (hash) {
+      return hash
+    }
+
+    hash = await this._db.get(key)
+    this._cache.numberToHash.set(key, hash)
+
+    return hash
+  }
+
+  async _getTd (hash, number) {
+    const key = tdKey(number, hash)
+    let td = this._cache.td.get(key)
+    if (td) {
+      return new BN(rlp.decode(td))
+    }
+
+    td = await this._db.get(key)
+    this._cache.td.set(key, td)
+
+    return new BN(rlp.decode(td))
+  }
+
+  async _getBody (hash, number) {
+    const key = bodyKey(number, hash)
+    let encodedBody = this._cache.body.get(key)
+    if (encodedBody) {
+      return rlp.decode(encodedBody)
+    }
+
+    encodedBody = await this._db.get(key)
+    this._cache.body.set(key, encodedBody)
+
+    return rlp.decode(encodedBody)
+  }
+
```

Comment:
A suggestion for DRY/readability:

```suggestion
  async _getBody (hash, number) {
    const key = bodyKey(number, hash)
    let encodedBody = this._cache.body.get(key)
    if (!encodedBody) {
      encodedBody = await this._db.get(key)
      this._cache.body.set(key, encodedBody)
    }
    return rlp.decode(encodedBody)
  }
```